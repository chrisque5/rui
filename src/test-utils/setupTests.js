// The below part of the code is to fix the issue
// TypeError: window.matchMedia is not a function
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});
