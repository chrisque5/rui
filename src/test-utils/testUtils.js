/* eslint-disable import/no-extraneous-dependencies, react/prop-types */
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Form } from 'antd';

const middlewares = [thunk];
const renderWithForm = (
  ui,
  {
    initialState,
    store = configureMockStore(middlewares)(initialState),
    ...renderOptions
  } = {},
) => {
  const wrapper = ({ children }) => <Provider store={store}><Form>{children}</Form></Provider>;

  return render(ui, { wrapper, ...renderOptions });
};

const renderWithProvider = (
  ui,
  {
    initialState,
    store = configureMockStore(middlewares)(initialState),
    ...renderOptions
  } = {},
) => {
  const wrapper = ({ children }) => <Provider store={store}>{children}</Provider>;

  return render(ui, { wrapper, ...renderOptions });
};
// re-export everything
export * from '@testing-library/react';
// override render method
export { renderWithForm, renderWithProvider };
