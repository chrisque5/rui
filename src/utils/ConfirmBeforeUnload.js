import { useEffect, useCallback } from 'react';

const ConfirmBeforeUnload = () => {
  const onBeforeUnload = useCallback((e) => {
    e.preventDefault();
    e.returnValue = '';
  }, []);

  useEffect(() => {
    window.addEventListener('beforeunload', onBeforeUnload);

    return () => {
      window.removeEventListener('beforeunload', onBeforeUnload);
    };
  }, [onBeforeUnload]);

  return null;
};

export default ConfirmBeforeUnload;
