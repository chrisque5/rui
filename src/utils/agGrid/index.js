import {
  approvalsCellRenderer, approvalsSummaryCellRenderer, isUnderInvestigationCellRenderer, stpCellRenderer, stpFilterCellRenderer,
} from './js-renderers';
import {
  dateFormatter, formatDate, numberFormatter, timeFormatter,
  valueDateFormatter, fixingDateFormatter, publishDateFormatter,
  priceFormatter, brokerageFormatter,
} from './formatters';
import { dateFilterComparator, numberFilterComparitor, timeFilterComparator } from './comparitors';
import { getExportMenuItems } from './exportUtils';

export {
  approvalsCellRenderer,
  approvalsSummaryCellRenderer,
  dateFilterComparator,
  dateFormatter,
  formatDate,
  getExportMenuItems,
  isUnderInvestigationCellRenderer,
  numberFilterComparitor,
  numberFormatter,
  stpCellRenderer,
  stpFilterCellRenderer,
  timeFilterComparator,
  timeFormatter,
  valueDateFormatter,
  fixingDateFormatter,
  publishDateFormatter,
  priceFormatter,
  brokerageFormatter,
};
