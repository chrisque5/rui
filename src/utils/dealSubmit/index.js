import mapFormToFwdChain from './mapFormToFwdChain';
import mapFormToNdfChain from './mapFormToNdfChain';
import mapFormToSptChain from './mapFormToSptChain';

export { mapFormToFwdChain, mapFormToNdfChain, mapFormToSptChain };
