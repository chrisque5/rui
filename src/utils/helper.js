import moment from 'moment';
import * as math from 'mathjs';
import { Modal } from 'antd';

export const isValidTerm = (termToTest, selections) => {
  if (termToTest.length < 2) return false;

  const lowerCaseTermToTest = termToTest.toLowerCase();

  const endsWithPeriod = (term) => {
    const periods = ['d', 'w', 'm', 'y'];
    return (periods.includes(term.charAt(term.length - 1)));
  };

  const firstCharacterIsNumber = (term) => {
    const result = term.substring(0, term.length - 1).match(/^[0-9]*$/);
    return result ? result.length > 0 : false;
  };

  const isStaticSelectionValid = (term) => {
    const validStaticTerms = selections.map((x) => x.toLowerCase());
    if (validStaticTerms.includes(term)) {
      return true;
    }

    if (term.startsWith('bmf') || term.startsWith('imm')) {
      let termDaysOnly = term.replace('bmf', '');
      termDaysOnly = termDaysOnly.replace('imm', '');
      const beforeConversion = termDaysOnly;
      termDaysOnly = parseInt(termDaysOnly, 10);
      if (beforeConversion.length === termDaysOnly.toString().length) {
        return (termDaysOnly % 1 === 0 && termDaysOnly > 0 && termDaysOnly < 101);
      }
    }
    return false;
  };

  return isStaticSelectionValid(lowerCaseTermToTest)
    || (endsWithPeriod(lowerCaseTermToTest) && firstCharacterIsNumber(lowerCaseTermToTest));
};

export const doesArrayExistWithValue = (array, index = 0) => {
  if (Array.isArray(array)) {
    if (array && array.length > 0 && array[index]) {
      return true;
    }
  }
  return false;
};

export const setTodayOnNull = (form, id, value) => {
  if (!value) {
    setTimeout(() => {
      form.setFieldsValue({ [id]: moment() });
    });
  }
};

export const dayDifferenceIgnoreHours = (earlierDateMoment, laterDateMoment) => {
  const date1 = earlierDateMoment.clone().startOf('day');
  const date2 = laterDateMoment.clone().startOf('day');
  const dayCount = date1.diff(date2, 'days');

  return dayCount;
};

export const getDateInLocalTZ = () => moment().tz(moment.tz.guess(true)).format('YYYY-MM-DD');
export const getDateTimeInLocalTZ = () => moment().tz(moment.tz.guess(true));

export const parseValueAsBigNumber = (value) => {
  try {
    return math.bignumber(value);
  } catch (err) {
    return null;
  }
};

export const formatAmount = (val) => {
  const inputValue = val.replace(/\s/g, '');
  let multiplier = 1;
  const oneM = 1000000;
  const oneB = 1000000000;
  let amt = inputValue;

  if (Number.isNaN(Number(inputValue)) && inputValue.length > 1) {
    const lastChar = inputValue.charAt(inputValue.length - 1).toLowerCase();
    amt = Number(inputValue.substring(0, inputValue.length - 1));
    if (lastChar === 'm') {
      multiplier = oneM;
    } else if (lastChar === 'b') {
      multiplier = oneB;
    } else {
      return inputValue;
    }
  } else if (Number(inputValue) < oneM) {
    multiplier = oneM;
  }

  const calculatedAmount = math.multiply(math.bignumber(amt), math.bignumber(multiplier));
  return calculatedAmount.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
};

export const isValidDate = (date, dateFormat) => moment(date, dateFormat || 'L', true).isValid();

export const findCurrencyPair = (ccy1, ccy2, currencies) => (
  currencies.find(
    (currency) => currency.baseCurrency === ccy1 && currency.counterCurrency === ccy2,
  )
);

export const filterSpotExecutionVenues = (execVenues) => execVenues.filter((i) => i.venueId === 'XOFF');

export const isDealTypeSame = (dealType1, dealType2) => (dealType1 === dealType2);

export const isStrategyTypeSame = (strategyType1, strategyType2) => (strategyType1 === strategyType2);

export const ConfirmUnsavedChanges = (message, onYes, onNo) => {
  Modal.confirm({
    title: 'Confirm',
    content: message,
    okText: 'Yes',
    cancelText: 'No',
    onOk: () => {
      if (onYes) {
        setTimeout(() => {
          onYes();
        });
      }
    },
    onCancel: () => {
      if (onNo) {
        setTimeout(() => {
          onNo();
        });
      }
    },
  });
};

export const isObjectEmpty = (obj) => Object.keys(obj).length === 0 && obj.constructor === Object;

export const getUniqueID = () => {
  const s4 = () => Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1);
  return `${s4() + s4()}-${s4()}`;
};

export const showPreviousValueToolTip = (currentValue, previousValue) => {
  if (currentValue !== previousValue) {
    return { title: `Previous Value: ${previousValue || 'N/A'}`, className: 'deal-input-dirty' };
  }
  return { title: '', className: '' };
};
