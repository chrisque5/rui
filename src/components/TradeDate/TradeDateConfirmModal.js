import { ExclamationCircleFilled } from '@ant-design/icons';
import { Button, Modal } from 'antd';
import { useDispatch } from 'react-redux';
import { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import { ids } from '../../utils/constants';
import { toggleTradeDateSubmitPopup } from '../../actions/uiActions';
import { getDateTimeInLocalTZ } from '../../utils/helper';

const TradeDateConfirmModal = ({
  form, isVisible, onAccept,
}) => {
  const dispatch = useDispatch();
  const [tradeDate, setTradeDate] = useState();
  const [tradeTime, setTradeTime] = useState();
  const [currentDate, setCurrentDate] = useState(getDateTimeInLocalTZ());

  useEffect(() => {
    if (isVisible) {
      setCurrentDate(getDateTimeInLocalTZ());
    }
  }, [isVisible]);

  useEffect(() => {
    const formTradeDate = form.getFieldValue(ids.TRADE_DATE);
    const formTradeTime = form.getFieldValue(ids.TRADE_TIME);

    if (formTradeDate) {
      setTradeDate(moment(formTradeDate).format('Do MMMM YYYY'));
    }

    if (formTradeTime) {
      setTradeTime(moment(formTradeTime).format('HH:mm:ss'));
    }
  }, [form, isVisible]);

  const closePopup = () => {
    dispatch(toggleTradeDateSubmitPopup(false));
  };

  const onclick = () => {
    closePopup();
    onAccept();
  };

  return (
    <Modal
      className="confirm-modal"
      title="Confirm Trade Date/Time"
      visible={isVisible}
      onCancel={closePopup}
      bodyStyle={{ padding: '10px' }}
      footer={[
        <Button data-testid="trade-date-modal-cancel" key="back" onClick={closePopup}>Cancel</Button>,
        <Button data-testid="trade-date-modal-save" key="submit" type="primary" onClick={onclick}>Accept</Button>,
      ]}
      destroyOnClose
    >
      <div className="modal-body">
        <span className="modal-header">
          <ExclamationCircleFilled />
          Trade Date/Time has been amended.
        </span>
        <span className="modal-times">
          <span>Current Trade Date/Time: </span>
          <span>{`${moment(currentDate).format('Do MMMM YYYY')} ${moment(currentDate).format('HH:mm:ss')}`}</span>
        </span>
        <span className="modal-times">
          <span>Amended Trade Date/Time:  </span>
          <span>{`${tradeDate} ${tradeTime}`}</span>
        </span>
      </div>
    </Modal>
  );
};

TradeDateConfirmModal.propTypes = {
  form: PropTypes.shape().isRequired,
  isVisible: PropTypes.bool.isRequired,
  onAccept: PropTypes.func.isRequired,
};

export default TradeDateConfirmModal;
