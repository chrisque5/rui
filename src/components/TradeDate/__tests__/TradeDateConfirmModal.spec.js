import { render, fireEvent, waitFor } from 'test-utils/testUtils';
import redux from 'react-redux';
import TradeDateConfirmModal from '../TradeDateConfirmModal';
import { ids } from '../../../utils/constants';

jest.mock('react-redux', () => ({ useDispatch: () => () => { }, useSelector: () => ({}) }));
jest.mock('../../../utils/helper', () => ({ getDateTimeInLocalTZ: jest.fn().mockReturnValue({ momentTz: 1586386800000 }) }));
jest.mock('moment', () => () => ({ format: () => '2099–01–01' }));

const dispatchSpy = jest.spyOn(redux, 'useDispatch');
afterEach(() => {
  dispatchSpy.mockClear();
});

const mockForm = {
  getFieldValue: (field) => {
    if (field === ids.TRADE_DATE || field === ids.TRADE_TIME) {
      return '01-01-2030';
    }
    return 'dummy';
  },
};

const onAccept = jest.fn();
const defaultProps = { form: mockForm, isVisible: true, onAccept };

test('<TradeDateConfirmModal /> renders snapshot as expected', () => {
  const { container } = render(<TradeDateConfirmModal {...defaultProps} />);
  expect(container).toMatchSnapshot();
});

test('<TradeDateConfirmModal /> renders snapshot as expected when no dates set in form', () => {
  const props = {
    form: { getFieldValue: () => {} },
    isVisible: true,
    onAccept,
  };

  const { container } = render(<TradeDateConfirmModal {...props} />);
  expect(container).toMatchSnapshot();
});

test.skip('<TradeDateConfirmModal /> calls dispatch and onAccept on clicking of accept button', async () => {
  const { getByTestId } = render(<TradeDateConfirmModal {...defaultProps} />);
  const acceptButton = await waitFor(() => getByTestId('trade-date-modal-save'));
  fireEvent.click(acceptButton);
  expect(defaultProps.onAccept).toHaveBeenCalled();
  expect(dispatchSpy).toHaveBeenCalledTimes(2);
});

test.skip('<TradeDateConfirmModal /> calls dispatch but not onAccept on clicking of cancel button', async () => {
  const { getByTestId } = render(<TradeDateConfirmModal {...defaultProps} />);
  const cancelButton = await waitFor(() => getByTestId('trade-date-modal-cancel'));
  fireEvent.click(cancelButton);
  expect(defaultProps.onAccept).not.toHaveBeenCalledTimes(41);
  expect(dispatchSpy).toHaveBeenCalledTimes(2);
});
