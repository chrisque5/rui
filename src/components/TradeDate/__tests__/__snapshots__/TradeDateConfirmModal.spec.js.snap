// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TradeDateConfirmModal /> renders snapshot as expected 1`] = `<div />`;

exports[`<TradeDateConfirmModal /> renders snapshot as expected when no dates set in form 1`] = `<div />`;
